Backend/starter_project_mavericks/BlogApp/BlogApp.Identity/Repositories/AuthRepository.cs

```diff
@@ -0,0 +1,146 @@
+        {
+            _userManager = userManager;
+            _signInManager = signInManager;
+            _mapper = mapper;
+            _configuration = configuration;
+            _context = context;
+        }
+
+        // Delete user
+        public async Task<bool> DeleteUserAsync(string userId)
+        {
+            var user = await _userManager.FindByIdAsync(userId);
+            if (user == null)
+            {
+                throw new NotFoundException(nameof(User), userId);
+            }
+            var result = await _userManager.DeleteAsync(user);
+            return result.Succeeded;
+        }        
+
+        public async Task<SignInResponse> SignInAsync(SigninFormDto signInFormDto)
+        {
+            
+                
+                    var userExist = await _userManager.FindByEmailAsync(signInFormDto.Email);
+                    if (userExist != null)
+                    {
+                        var result = await _signInManager.CheckPasswordSignInAsync(userExist, signInFormDto.Password, false);
+
+                        if (result.Succeeded)
+                        {
+                            var token = await GenerateToken(userExist);
+                            return new SignInResponse
+                            {
+                                Id = userExist.Id,
+                                Email = userExist.Email,
+                                FirstName = userExist.FirstName,
+                                LastName = userExist.LastName,
+                                Token = token.ToString()
+                            };
+                        }
+                    }
+                    throw new BadRequestException("Invalid Credential");
+
+            }
+            
+            
+        
+    
+
+        public async Task<SignUpResponse> SignUpAsync(SignupFormDto signUpFormDto)
+        {
+            using( var transaction = _context.Database.BeginTransaction())
+            {
+                try 
+                {
+                    var userExist = await _userManager.FindByEmailAsync(signUpFormDto.Email);
+                    if (userExist == null)
+                    {
+                        var user = _mapper.Map<User>(signUpFormDto);
+                        user.UserName = signUpFormDto.Email;
+                        var result = await _userManager.CreateAsync(user, signUpFormDto.Password);
+                        if (result.Succeeded)
+                        {
+                            await _userManager.AddToRoleAsync(user, "User");
+                            var createdUser = await _userManager.FindByNameAsync(user.UserName);
+                            var response = _mapper.Map<SignUpResponse>(createdUser);
+
+                            return response;
+                            await transaction.CommitAsync();
+                        }
+                        throw new BadRequestException("Email already exists");
```

Comment:
This thrown exception message should be relevant to why _userManager faild to create the user. There is no user by this email couse we have validated that at the start our logic.